### ISSUE TITLE
[HELP]: flutter_sound audio package conflicts

### ISSUE THREAD

**Issue by quickratio** [Canardoux/flutter_sound#855]
Hi

Hope you're well; iOS specific (problem doesnt occur on android), anytime i use another package that uses audio, flutter_sound will cause the following crash, is there any known workarounds? As soon as audio is called from any other package e.g. flutter_tts (i've tested several), while trying to openRecorder i will get this error and the app will immediately crash,

Thanks in advance!

```

*** Terminating app due to uncaught exception 'com.apple.coreaudio.avfaudio', reason: 'required condition is false: IsFormatSampleRateAndChannelCountValid(format)'
*** First throw call stack:
(0x181002d3c 0x19837c6a8 0x1810cc5bc 0x1dca7eea0 0x1dcaec9a0 0x1dcad4358 0x1dcb57224 0x1dcb3b828 0x1036a4464 0x1036a4c84 0x1036a32b8 0x10258f384 0x1025907f8 0x10a36a75c 0x109ea3cf8 0x10a231bb4 0x10a1591c0 0x10a15cf24 0x18101706c 0x180fa13d8 0x180f9c06c 0x180f7bf44 0x180f8ee1c 0x1a13399a0 0x1837c2b90 0x18355816c 0x102313a20 0x102c8c250)
libc++abi: terminating with uncaught exception of type NSException
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x00000001bb1369e8 libsystem_kernel.dylib`__pthread_kill + 8
libsystem_kernel.dylib`__pthread_kill:
->  0x1bb1369e8 <+8>:  b.lo   0x1bb136a04               ; <+36>
    0x1bb1369ec <+12>: stp    x29, x30, [sp, #-0x10]!
    0x1bb1369f0 <+16>: mov    x29, sp
    0x1bb1369f4 <+20>: bl     0x1bb132670               ; cerror_nocancel
Target 0: (Runner) stopped.
Lost connection to device.

```

**Comment by Larpoux** [comment]
Which Flutter Sound version do you use? 8.x or 9.x?
9.x is supposed to be compatible with others libraries.

**Comment by quickratio** [comment]
Hi Larpoux, thanks for your response,

flutter_sound: ^9.1.3, the latest

It appears to be an issue with other mic stream packages too, basically if i play a sound from any other package the recorder will stop, i suspect that apple must have made a breaking change as it was working perfectly fine before december 2021, im happy to pay you to help me on this one! ive been stuck for months! haha.

**Comment by quickratio** [comment]
Here is a sample widget,

It is a modification of your stream script, adapted to use google speech, i have removed the google speech part as it wont work without credentials, but it still demonstrates the crash that happens.

You will need to additionally reference the TTS package in pubspec:  flutter_tts: ^3.3.3

Steps to reproduce:
1. Press record button (should load without issue)
2. Press the "Say hello" button
3. App will crash and get the same error as reference above as soon as it tries to run _mRecorder.startRecorder

```
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter_sound/flutter_sound.dart';
import 'package:flutter_tts/flutter_tts.dart';

import 'package:permission_handler/permission_handler.dart';
import 'package:rxdart/rxdart.dart';

const int tSampleRate = 16000;
typedef _Fn = void Function();

/// Example app.
class RecordToStreamExample extends StatefulWidget {
  @override
  _RecordToStreamExampleState createState() => _RecordToStreamExampleState();
}

class _RecordToStreamExampleState extends State<RecordToStreamExample> {
  FlutterSoundPlayer _mPlayer = FlutterSoundPlayer();
  FlutterSoundRecorder _mRecorder = FlutterSoundRecorder();
  bool _mPlayerIsInited = false;
  bool _mRecorderIsInited = false;
  bool _mplaybackReady = false;
  String _mPath;
  StreamSubscription _mRecordingDataSubscription;

  Future<void> _openRecorder() async {
    var status = await Permission.microphone.request();
    if (status != PermissionStatus.granted) {
      throw RecordingPermissionException('Microphone permission not granted');
    }
    await _mRecorder.openRecorder();
    setState(() {
      _mRecorderIsInited = true;
    });
  }

  @override
  void initState() {
    super.initState();
    // Be careful : openAudioSession return a Future.
    // Do not access your FlutterSoundPlayer or FlutterSoundRecorder before the completion of the Future
    _mPlayer.openPlayer().then((value) {
      setState(() {
        _mPlayerIsInited = true;
      });
    });
    _openRecorder();
  }

  @override
  void dispose() {
    stopPlayer();
    _mPlayer.closePlayer();
    _mPlayer = null;

    stopRecorder();
    _mRecorder.closeRecorder();
    _mRecorder = null;
    super.dispose();
  }

  FlutterTts flutterTts = FlutterTts();

  Future vocalCardTTS() async {
    await flutterTts.setIosAudioCategory(IosTextToSpeechAudioCategory.ambient,
        [
          IosTextToSpeechAudioCategoryOptions.allowBluetooth,
          IosTextToSpeechAudioCategoryOptions.allowBluetoothA2DP,
          IosTextToSpeechAudioCategoryOptions.mixWithOthers
        ],
        IosTextToSpeechAudioMode.voicePrompt
    );
    await flutterTts.awaitSpeakCompletion(true);
    await flutterTts.setSharedInstance(true);
    await flutterTts.setLanguage('en_US');
    await flutterTts.setSpeechRate(0.5);
    await flutterTts.speak('Hello');
  }

  // ----------------------  Here is the code to record to a Stream ------------

  StreamSubscription<List<int>> audioStreamSubscription;
  BehaviorSubject<List<int>> audioStream;

  Future<void> record() async {
    assert(_mRecorderIsInited && _mPlayer.isStopped);
    var recordingDataController = StreamController<Food>();

    audioStream = BehaviorSubject<List<int>>();
    _mRecordingDataSubscription = recordingDataController.stream.listen((buffer) {
      if (buffer is FoodData) {
        audioStream.add(buffer.data);
      }
    });

    await _mRecorder.startRecorder(
      toStream: recordingDataController.sink,
      codec: Codec.pcm16,
      numChannels: 1,
      sampleRate: tSampleRate,
    );

    setState(() {});
  }

  Future<void> stopRecorder() async {
    await _mRecorder.stopRecorder();
    if (_mRecordingDataSubscription != null) {
      await _mRecordingDataSubscription.cancel();
      _mRecordingDataSubscription = null;
    }
    _mplaybackReady = true;
  }

  _Fn getRecorderFn() {
    if (!_mRecorderIsInited || !_mPlayer.isStopped) {
      return null;
    }
    return _mRecorder.isStopped
        ? record
        : () {
      stopRecorder().then((value) => setState(() {}));
    };
  }

  void play() async {
    assert(_mPlayerIsInited &&
        _mplaybackReady &&
        _mRecorder.isStopped &&
        _mPlayer.isStopped);
    await _mPlayer.startPlayer(
        fromURI: _mPath,
        sampleRate: tSampleRate,
        codec: Codec.pcm16,
        numChannels: 1,
        whenFinished: () {
          setState(() {});
        }); // The readability of Dart is very special :-(
    setState(() {});
  }

  Future<void> stopPlayer() async {
    await _mPlayer.stopPlayer();
  }

  _Fn getPlaybackFn() {
    if (!_mPlayerIsInited || !_mplaybackReady || !_mRecorder.isStopped) {
      return null;
    }
    return _mPlayer.isStopped
        ? play
        : () {
      stopPlayer().then((value) => setState(() {}));
    };
  }

  // ----------------------------------------------------------------------------------------------------------------------

  @override
  Widget build(BuildContext context) {
    Widget makeBody() {
      return Column(
        children: [
          Container(
            margin: const EdgeInsets.all(3),
            padding: const EdgeInsets.all(3),
            height: 80,
            width: double.infinity,
            alignment: Alignment.center,
            decoration: BoxDecoration(
              color: Color(0xFFFAF0E6),
              border: Border.all(
                color: Colors.indigo,
                width: 3,
              ),
            ),
            child: Row(children: [
              ElevatedButton(
                onPressed: getRecorderFn(),
                //color: Colors.white,
                //disabledColor: Colors.grey,
                child: Text(_mRecorder.isRecording ? 'Stop' : 'Record'),
              ),
              SizedBox(
                width: 20,
              ),
              Text(_mRecorder.isRecording
                  ? 'Recording in progress'
                  : 'Recorder is stopped'),
            ]),
          ),
          Container(
            margin: const EdgeInsets.all(3),
            padding: const EdgeInsets.all(3),
            height: 80,
            width: double.infinity,
            alignment: Alignment.center,
            decoration: BoxDecoration(
              color: Color(0xFFFAF0E6),
              border: Border.all(
                color: Colors.indigo,
                width: 3,
              ),
            ),
            child: Row(children: [
              ElevatedButton(
                onPressed: vocalCardTTS,
                //color: Colors.white,
                //disabledColor: Colors.grey,
                child: Text('Say hello'),
              ),
              SizedBox(
                width: 20,
              ),
              Text('Will talk using text to speech'),
            ]),
          ),
        ],
      );
    }

    return Scaffold(
      backgroundColor: Colors.blue,
      body: makeBody(),
    );
  }
}
```

**Comment by quickratio** [comment]
haha no please keep it, no obligation, take it as a thank you so far, if you can help, then bonus. Maybe when you have time i can go over my project with you, right now im focusing on launching, and this blocked me from launchin, but im a one man team, so no big deal :)

**Comment by Larpoux** [comment]
@quickratio : 

I cut and pasted your code, and I get also a crash when launched.
I am not sure it is same than your crash, but it seems that something not good.

I am going to work on that.

**Comment by quickratio** [comment]
Thanks Larpoux, from my testing these are the conditions for crash:

1. Recorder is initiated
2. Any audio is played from any package other than flutter_sound (if i use flutter sound only it works fine)
3. Try to initiate recorder will cause a crash even after a hot reload, i've never had a crash on building an app, which is unusually for that to happen to you.

**Comment by Larpoux** [comment]
My crash was just because I forgot to put Recorder permission inside Runner/info.plist
```
	<key>NSMicrophoneUsageDescription</key>
	<string>This example wants to access your microphone</string>
	<key>UIBackgroundModes</key>
	<array>
		<string>audio</string>
		<string>processing</string>
	</array>
```

I am going to try to reproduce your crash, now.

**Comment by Larpoux** [comment]
@quickratio : 

Can you download [this project](), and run the flutter App : toto/titi,
and tell me how to have the crash.

**Comment by quickratio** [comment]
Trying it now :)

**Comment by Larpoux** [comment]
OK : I have it :)

on a real iPHONE.

Do not try my example. I have enough to work now.
Talk with you later.

**Comment by quickratio** [comment]
Thanks Larpoux you're the best! :) i actually havent tried on an emulator, but yes i test on real physical phones, as sometimes in the past testing on emulators doesnt mean it works on a physical device :)

**Comment by Larpoux** [comment]
Just to let you know that I progress. Tomorrow I will try a fix.
I will explain to you later.

**Comment by Larpoux** [comment]
@quickratio : 

### 1. TLDR : Restore a correctAudio Session before starting the recorder

A correct fix for your issue is : 
- Add a dependency to the [Audio-Session plugin](https://pub.dev/packages/audio_session)
- Set a correct audio session for recording before startRecorder() : 
```dart

    final session = await AudioSession.instance;
    await session.configure(AudioSessionConfiguration(
      avAudioSessionCategory: AVAudioSessionCategory.playAndRecord,
      avAudioSessionCategoryOptions: AVAudioSessionCategoryOptions.allowBluetooth,
      avAudioSessionMode: AVAudioSessionMode.spokenAudio,
      avAudioSessionRouteSharingPolicy: AVAudioSessionRouteSharingPolicy.defaultPolicy,
      avAudioSessionSetActiveOptions: AVAudioSessionSetActiveOptions.none,
      androidAudioAttributes: const AndroidAudioAttributes(
        contentType: AndroidAudioContentType.speech,
        flags: AndroidAudioFlags.none,
        usage: AndroidAudioUsage.voiceCommunication,
      ),
      androidAudioFocusGainType: AndroidAudioFocusGainType.gain,
      androidWillPauseWhenDucked: true,
    ));
```

### 2. Long version

On iOS, each App has its own Audio Session, but unfortunately, there is just one audio-session for each running App. iOS is just a piece of bullshit. Because of this uniqueness, we get many compatibility issues when an App wants to use several packages which does not cooperate.

On Flutter Sound 8.x, we tried to manage ourself the Audio Sessions. It was convenient and was fine most of the time. But because of this management, Flutter Sound was badly compatible with others plugin.
What happened was that each Plugin tried to fight for doing Session management.

The good way to fix this incompatibility, was to remove from Flutter Sound everything about Audio Session management, and rely on a federating manager (e.g. Audio_session). It is what we did in FS 9.x

Everything will work correctly if no Audio plugins do Audio Session management, and if everybody rely on the federator : Audio Session.
Actually I know two plugin which are good citizen and do not play with audio sessions
- Flutter Sound 9.x
- Just Audio

But many others plugin continue to try to manage the audio-sessions. To solve that, you will need to to restore a correct audio session before startRecorder() (i.e. a session correct for doing a recording).

A crash is a terrible thing and is not good. A fix could be that FS checks that the session is correct and set itself a correct Session if necessary. But we will have again the same problems as in 8.x.

Another fix could be to check that the session is correct and throw an exception if not.
I think that it is would be good and that is want I am going to do in 9.8

**Comment by quickratio** [comment]
awesome thanks larpoux! will try as soon as i can!

**Comment by Larpoux** [comment]
Hi my friend.
Of course you may send me a notice when your App will be launched. I just hope that this App wiil be free and Open Source 😉 .

My email addr is larpoux@gmail.com. You can also join me on [discord](https://discord.gg/MWNfvaTHmQ) if this is related to Flutter Sound or Taudio.

I will never forget that you bought me 20 coffees !!! Of course, do not hesitate to fill a new issue if you have other problems.

Actually I am busy working on a new plugin doing Audio Streams on Flutter and on other frameworks (Blazor, Microsoft MAUI, React Native, ...).
The goal is to offer an API compatibile with [Web Audio API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Audio_API) on : 
- iOS
- Android
- MacOS
- Linux (of course)
- And maybe Windows

The W3C API recommendation defines about 30 classes, each one having a dozen of methods.
And we must support that on many platforms and on many frameworks.
I hope to be able to release something later this year, or at least beginning of next year.

I am completely sure that this project will be very important for people wanting to do Audio Streams.
And the compliance with W3C Recommendation is something really good.
The plugin is not very complicate, but there are many, many things to work on.

Perhaps you will be one of the first Taudio user 😄  .
But it seems that you do not need important functionalities with your Streams.
If I understand correctly you only need a correct Stream to feed a Speach to Text Engine.

What I am interested by, is how you convert the Byte Arrays to an INT16 arrays. Several Flutter Sound users was not able to manage correctly the Little-Endian format and they will be interested by your work. And some of them have also problems with Base-50 encoding.
Everything you can share with other Flutter Sound users will be really great.

I love you @quickratio . Good luck for your own App.

**Comment by tan-yaka** [comment]
Hi @quickratio any chance you could share your conversion method so the output can be consumed by a STT service? I've tried many approaches with no success. I'm trying this with MS Speech Services btw

**Comment by quickratio** [comment]
@tan-yaka here is the full working script adapted for google speech package (third party flutter library), i've redacted the google credentials in the tags <google credentials>

I havent used azure stt (i find most azure products to be subpar), you'll need to set up a google speech service if you want to test this script, further instructions thereL https://pub.dev/packages/google_speech

```

import 'dart:async';
import 'dart:io';
import 'package:audio_session/audio_session.dart';
import 'package:flashwords/controllers/text_to_speech_controller.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_sound/flutter_sound.dart';
import 'package:flutter_tts/flutter_tts.dart';
import 'package:get/get.dart';
import 'package:path_provider/path_provider.dart';
import 'package:permission_handler/permission_handler.dart';
import 'package:rxdart/rxdart.dart';
import 'package:google_speech/google_speech.dart';

const int tSampleRate = 16000;
typedef _Fn = void Function();

/// Example app.
class RecordToStreamExample extends StatefulWidget {
  @override
  _RecordToStreamExampleState createState() => _RecordToStreamExampleState();
}

class _RecordToStreamExampleState extends State<RecordToStreamExample> {
  FlutterSoundPlayer _mPlayer = FlutterSoundPlayer();
  FlutterSoundRecorder _mRecorder = FlutterSoundRecorder();
  bool _mPlayerIsInited = false;
  bool _mRecorderIsInited = false;
  bool _mplaybackReady = false;
  String _mPath;
  StreamSubscription _mRecordingDataSubscription;

  Future<void> _openRecorder() async {
    var status = await Permission.microphone.request();
    if (status != PermissionStatus.granted) {
      throw RecordingPermissionException('Microphone permission not granted');
    }
    await _mRecorder.openRecorder();
    setState(() {
      _mRecorderIsInited = true;
    });
  }

  Future<void> setAudioConfig() async {
    final session = await AudioSession.instance;
    await session.configure(AudioSessionConfiguration(
      avAudioSessionCategory: AVAudioSessionCategory.playAndRecord,
      avAudioSessionCategoryOptions: AVAudioSessionCategoryOptions.allowBluetooth,
      avAudioSessionMode: AVAudioSessionMode.spokenAudio,
      avAudioSessionRouteSharingPolicy: AVAudioSessionRouteSharingPolicy.defaultPolicy,
      avAudioSessionSetActiveOptions: AVAudioSessionSetActiveOptions.none,
      androidAudioAttributes: const AndroidAudioAttributes(
        contentType: AndroidAudioContentType.speech,
        flags: AndroidAudioFlags.none,
        usage: AndroidAudioUsage.voiceCommunication,
      ),
      androidAudioFocusGainType: AndroidAudioFocusGainType.gain,
      androidWillPauseWhenDucked: true,
    ));
  }

  @override
  void initState() {
    super.initState();
    // Be careful : openAudioSession return a Future.
    // Do not access your FlutterSoundPlayer or FlutterSoundRecorder before the completion of the Future
    _mPlayer.openPlayer().then((value) {
      setState(() {
        _mPlayerIsInited = true;
      });
    });
    _openRecorder();
  }

  @override
  void dispose() {
    stopPlayer();
    _mPlayer.closePlayer();
    _mPlayer = null;

    stopRecorder();
    _mRecorder.closeRecorder();
    _mRecorder = null;
    super.dispose();
  }

  FlutterTts flutterTts = FlutterTts();

  Future vocalCardTTS() async {
    await flutterTts.setIosAudioCategory(IosTextToSpeechAudioCategory.ambient,
        [
          IosTextToSpeechAudioCategoryOptions.allowBluetooth,
          IosTextToSpeechAudioCategoryOptions.allowBluetoothA2DP,
          IosTextToSpeechAudioCategoryOptions.mixWithOthers
        ],
        IosTextToSpeechAudioMode.voicePrompt
    );
    await flutterTts.awaitSpeakCompletion(true);
    await flutterTts.setSharedInstance(true);
    await flutterTts.setLanguage('en_US');
    await flutterTts.setSpeechRate(0.5);
    await flutterTts.speak('Hello');
  }

  Future<IOSink> createFile() async {
    var tempDir = await getTemporaryDirectory();
    _mPath = '${tempDir.path}/flutter_sound_example.pcm';
    var outputFile = File(_mPath);
    if (outputFile.existsSync()) {
      await outputFile.delete();
    }
    return outputFile.openWrite();
  }

  // ----------------------  Here is the code to record to a Stream ------------

  StreamSubscription<List<int>> audioStreamSubscription;
  BehaviorSubject<List<int>> audioStream;

  Future<void> record() async {
    assert(_mRecorderIsInited && _mPlayer.isStopped);
    var recordingDataController = StreamController<Food>();

    audioStream = BehaviorSubject<List<int>>();
    _mRecordingDataSubscription = recordingDataController.stream.listen((buffer) {
      if (buffer is FoodData) {
        audioStream.add(buffer.data);
      }
    });

    await setAudioConfig().then((value) async {
      await _mRecorder.startRecorder(
        toStream: recordingDataController.sink,
        codec: Codec.pcm16,
        numChannels: 1,
        sampleRate: tSampleRate,
      );
    });

    final serviceAccount = ServiceAccount.fromString(
        '${(await rootBundle.loadString('assets/<YOUR JSON GOOGLE CREDENTIAL FILE HERE>.json'))}');
    final speechToText = SpeechToText.viaServiceAccount(serviceAccount);
    final config = _getConfig();

    final responseStream = speechToText.streamingRecognize(
        StreamingRecognitionConfig(config: config, interimResults: true),
        audioStream);

    responseStream.listen((data) async {
      String text = data.results.map((e) => e.alternatives.first.transcript).join('\n');
      print("Google translate results: " + text);
    }, onDone: () async {

    });

    setState(() {});
  }

  RecognitionConfig _getConfig() => RecognitionConfig(
    encoding: AudioEncoding.LINEAR16,
    model: RecognitionModel.basic,
    enableAutomaticPunctuation: false,
    sampleRateHertz: 16000,
    languageCode: 'es-ES',
    // languageCode: MapData.mapIndexToGoogleSTT[global.langTarget]
  );
  // --------------------- (it was very simple, wasn't it ?) -------------------

  Future<void> stopRecorder() async {
    await _mRecorder.stopRecorder();
    if (_mRecordingDataSubscription != null) {
      await _mRecordingDataSubscription.cancel();
      _mRecordingDataSubscription = null;
    }
    _mplaybackReady = true;
  }

  _Fn getRecorderFn() {
    if (!_mRecorderIsInited || !_mPlayer.isStopped) {
      return null;
    }
    return _mRecorder.isStopped
        ? record
        : () {
      stopRecorder().then((value) => setState(() {}));
    };
  }

  void play() async {
    assert(_mPlayerIsInited &&
        _mplaybackReady &&
        _mRecorder.isStopped &&
        _mPlayer.isStopped);
    await _mPlayer.startPlayer(
        fromURI: _mPath,
        sampleRate: tSampleRate,
        codec: Codec.pcm16,
        numChannels: 1,
        whenFinished: () {
          setState(() {});
        }); // The readability of Dart is very special :-(
    setState(() {});
  }

  Future<void> stopPlayer() async {
    await _mPlayer.stopPlayer();
  }

  _Fn getPlaybackFn() {
    if (!_mPlayerIsInited || !_mplaybackReady || !_mRecorder.isStopped) {
      return null;
    }
    return _mPlayer.isStopped
        ? play
        : () {
      stopPlayer().then((value) => setState(() {}));
    };
  }

  // ----------------------------------------------------------------------------------------------------------------------

  @override
  Widget build(BuildContext context) {
    Widget makeBody() {
      return Column(
        children: [
          Container(
            margin: const EdgeInsets.all(3),
            padding: const EdgeInsets.all(3),
            height: 80,
            width: double.infinity,
            alignment: Alignment.center,
            decoration: BoxDecoration(
              color: Color(0xFFFAF0E6),
              border: Border.all(
                color: Colors.indigo,
                width: 3,
              ),
            ),
            child: Row(children: [
              ElevatedButton(
                onPressed: getRecorderFn(),
                //color: Colors.white,
                //disabledColor: Colors.grey,
                child: Text(_mRecorder.isRecording ? 'Stop' : 'Record'),
              ),
              SizedBox(
                width: 20,
              ),
              Text(_mRecorder.isRecording
                  ? 'Recording in progress'
                  : 'Recorder is stopped'),
            ]),
          ),
          Container(
            margin: const EdgeInsets.all(3),
            padding: const EdgeInsets.all(3),
            height: 80,
            width: double.infinity,
            alignment: Alignment.center,
            decoration: BoxDecoration(
              color: Color(0xFFFAF0E6),
              border: Border.all(
                color: Colors.indigo,
                width: 3,
              ),
            ),
            child: Row(children: [
              ElevatedButton(
                onPressed: vocalCardTTS,
                //color: Colors.white,
                //disabledColor: Colors.grey,
                child: Text('Say hello'),
              ),
              SizedBox(
                width: 20,
              ),
              Text('Will talk using text to speech'),
            ]),
          ),
        ],
      );
    }

    return Scaffold(
      backgroundColor: Colors.blue,
      appBar: AppBar(
        leading: IconButton(
          icon: Icon(Icons.arrow_back),
          onPressed: () => Get.offAndToNamed(Get.previousRoute),
        ),
        title: const Text('Record to Stream ex.'),
      ),
      body: makeBody(),
    );
  }
}

```

**Comment by tan-yaka** [comment]
@quickratio many thanks for taking the time to reply. I am doing essentially the same as you except where you `        audioStream.add(buffer.data)` I'm sending over a websocket: `_socket.emit('stt-data', buffer.data)`. The problem I'm finding is the data in the buffer isn't of a format the STT service is expecting. From the MS docs the service expects data of the following format:
```
- PCM format
- One channel
- 16 bits per sample, 8,000 or 16,000 samples per second (16,000 bytes or 32,000 bytes per second)
- Two-block aligned (16 bit including padding for a sample)
```
It's the last requirement that's tripping me up and I'm not sure exactly what it means and how to go from a Uint8list to it. If you have any thoughts it would be much appreciated.

**Comment by Larpoux** [comment]
Hi guys and girls.
I have very serious family problems : actually I can't work on my computers.
And I can't do at the same time the necessary Flutter Sound maintenance and the development of Taudio.

I have some doubts that I will be able to release the first version of Taudio before the end of the year. 😕

**Comment by tan-yaka** [comment]
> @tan-yaka theyre the exact same requirements as google, have you tried adapting the code to test it?

@quickratio I'm sure your code works however my case is different in that you plug the BehaviorSubject instance straight into the Google speech recogniser whereas I am trying to send the data over a websocket to a Node service that proxies the STT service. I've done some more digging and the expected javascript data type is ArrayBuffer which what I'm trying to coerse the Uint8list into now. The service works fine with non Flutter client btw

**Comment by tan-yaka** [comment]
> Hi guys and girls. I have very serious family problems : actually I can't work on my computers. And I can't do at the same time the necessary Flutter Sound maintenance and the development of Taudio.
> 
> I have some doubts that I will be able to release the first version of Taudio before the end of the year. 😕

@Larpoux sorry to hear that. I'm new to flutter_sound but can see the amount of effort you've put into it. Hope things get better for you

**Comment by quickratio** [comment]
@Larpoux sorry to hear that, let me know if theres anything i can do to help?

**Comment by Larpoux** [comment]
@tan-yaka, @quickratio, others,

you are my friends, my only friends. What happens to my mother is horrible.
She is 94y old.

I feel better when I know that I have friends all around the earth

**Comment by tan-yaka** [comment]
@quickratio fyi I solved this by wrapping the data received by the server in a javascript Uint8Array which made the data consumable by the STT service. I couldn't get Flutter to output the data in a suitable format

**Comment by quickratio** [comment]
@tan-yaka great to hear you found a solution! Out of curiosity you're running the STT service on your own box?

**Comment by tan-yaka** [comment]
@quickratio yes the service is running locally for testing and proxies the MS STT service in the Azure cloud

**Comment by quickratio** [comment]
@tan-yaka interesting, may i ask why you chose Azure? from my testing it wasnt as accurate as google, interestingly the google STT api is superior to the local (on device) STT for either apple or android from my testing, (about 20 different languages i've tested).

**Comment by tan-yaka** [comment]
@quickratio we had a requirement to recognise some weird drug names (most of which I couldn't even pronounce!) and at the time MS STT was the only service that allowed us to train the speech model to recognise them using actual audio + transcriptions, and it worked really well. I know some of the other STT services now allow for this kind of training but we have invested a lot of time in perfecting our training method and amassed a lot of training data. I agree the Google STT API is far superior to the on device service. I've looked at other services like AWS Transcribe and found that to be the most difficult to implement and poorest performing in terms of accuracy. I'm going to be looking at some the smaller outfits offering STT like Speechly and Deepgram soon.
